src/VisualStudio/CSharp/Impl/CSharpPackage.cs

```diff
@@ -93,14 +100,19 @@ protected override VisualStudioWorkspaceImpl CreateWorkspace()
 
         protected override void Dispose(bool disposing)
         {
-            UnregisterObjectBrowserLibraryManager();
+            if (disposing)
```

Comment:
Is fire-and-forget the right thing to be doing here? What happens if our fire-and-forget is interleaving the shutting down of the service?

"Using services in Dispose()" has always been fraught with peril (and it looks like we weren't handling this great as it was here), but it seems we're introducing new peril here. What's the advice? What happens if we just don't unregister anything and delete this code? :smile:

Reply:
(oops: confused Task.Run() and JoinableTaskFactory.Run()) here

Reply:
This doesn't look like fire and forget to me. If anything throws an exception, it will propagate to Dispose's caller. JTF.Run does not eat exceptions.

---
In reply to: [192223953](https://github.com/dotnet/roslyn/pull/26787#discussion_r192223953) [](ancestors = 192223953)